src/System.Runtime.Serialization.Formatters/tests/SerializationGuardTests.cs

```diff
@@ -0,0 +1,178 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Reflection;
+using System.Runtime.Serialization.Formatters.Binary;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.Runtime.Serialization.Formatters.Tests
+{
+    public class SerializationGuardTests
+    {
+        [Fact]
+        void BlockAssemblyLoads()
+        {
+            TryPayload(new AssemblyLoader());
+        }
+
+        [Fact]
+        void BlockProcessStarts()
+        {
+            TryPayload(new ProcessStarter());
+        }
+
+        [Fact]
+        void BlockFileWrites()
+        {
+            TryPayload(new FileWriter());
+        }
+
+        [Fact]
+        void BlockReflectionDodging()
+        {
+            MemoryStream ms = new MemoryStream();
+            BinaryFormatter writer = new BinaryFormatter();
+            writer.Serialize(ms, new ReflectionDodger());
+            ms.Position = 0;
+
+            BinaryFormatter reader = new BinaryFormatter();
+            Assert.Throws<TargetInvocationException>(() => reader.Deserialize(ms));
+        }
+
+        [Fact]
+        void BlockAsyncDodging()
+        {
+            TryPayload(new AsyncDodger());
+        }
+
+        static void TryPayload(object payload)
+        {
+            MemoryStream ms = new MemoryStream();
+            BinaryFormatter writer = new BinaryFormatter();
+            writer.Serialize(ms, payload);
+            ms.Position = 0;
+
+            BinaryFormatter reader = new BinaryFormatter();
+            Assert.Throws<DeserializationBlockedException>(() => reader.Deserialize(ms)); // This should really look for DeserializationBlockedException, but it's not in all of the frameworks this test builds for
```

Comment:
Can we check for the correct exception type when testing on runtime where it is present?

Reply:
To verify serialization didn't fail for unrelated reason